### PR TITLE
Selective early media

### PR DESCRIPTION
**Pull request by cspiel1**
For an incoming call allow to specify if we want only
- early incoming audio,
- early incoming video,
- or audio+video in both directions (like before)

Done also: Ring tone and ringback tone is played as long no audio is in Session Progress SDP.

The ringback tone in this case depends on libre PR: https://github.com/baresip/re/pull/99

### CODE REVIEW

**Comment by juha-h** [comment]
How about `PRACK` to `183 Session Progress`?

**Comment by cspiel1** [comment]
> How about `PRACK` to `183 Session Progress`?
We had no demands so far.

**Comment by cspiel1** [comment]
Set this to draft because early video display don't arises.

File: include/baresip.h [hunk 1]
```diff
@@ -52,7 +52,9 @@ struct vidsz;
 enum answermode {
 	ANSWERMODE_MANUAL = 0,
 	ANSWERMODE_EARLY,
-	ANSWERMODE_AUTO
+	ANSWERMODE_AUTO,
+	ANSWERMODE_EARLY_VIDEO,
+	ANSWERMODE_EARLY_AUDIO,
```
Comment (alfredh) [inline include/baresip.h:57]: can you switch the order of "video" and "audio" ?

just like the SDP the audio always comes first :)

**Comment by alfredh** [comment]
there is a segfault in the testcode:

```
[ RUN      ] test_call_progress

Program received signal SIGSEGV, Segmentation fault.
call_progress (call=0x5555555f2d80) at src/call.c:1091
1091		call->ansvdir = video_strm(call_video(call))->ldir;
(gdb) bt          
#0  call_progress (call=0x5555555f2d80) at src/call.c:1091
#1  0x000055555556340c in event_handler (ua=0x5555555f12a0, ev=UA_EVENT_CALL_INCOMING, 
    call=0x5555555f2d80, prm=0x7fffffffd4b0 "sip:a@127.0.0.1", arg=0x7fffffffe000) at test/call.c:187
#2  0x000055555558b4c4 in ua_event (ua=0x5555555f12a0, ev=UA_EVENT_CALL_INCOMING, 
    call=0x5555555f2d80, fmt=0x5555556007f0 "sip:a@127.0.0.1") at src/event.c:303
#3  0x000055555559490d in call_event_handler (call=0x5555555f2d80, ev=CALL_EVENT_INCOMING, 
    str=0x7fffffffd700 "sip:a@127.0.0.1", arg=0x5555555f12a0) at src/ua.c:614
#4  0x0000555555581249 in call_event_handler (call=0x5555555f2d80, ev=CALL_EVENT_INCOMING, 
    fmt=0x5555556007f0 "sip:a@127.0.0.1") at src/call.c:235
#5  0x0000555555584c6f in call_accept (call=0x5555555f2d80, sess_sock=0x5555556276c0, 
    msg=0x555555611aa0) at src/call.c:1887
#6  0x0000555555596d8f in sipsess_conn_handler (msg=0x555555611aa0, arg=0x0) at src/ua.c:1780
#7  0x00007ffff7f4b965 in invite_handler (sock=0x5555556276c0, msg=0x555555611aa0)
    at src/sipsess/listen.c:244
#8  0x00007ffff7f4b9d9 in request_handler (msg=0x555555611aa0, arg=0x5555556276c0)
    at src/sipsess/listen.c:257
#9  0x00007ffff7f416b9 in sip_recv (sip=0x5555555efec0, msg=0x555555611aa0, start=0)
    at src/sip/transp.c:292
#10 0x00007ffff7f4198e in udp_recv_handler (src=0x7fffffffdce0, mb=0x55555564d7d0, arg=0x5555555f0aa0)
    at src/sip/transp.c:365
#11 0x00007ffff7f74d30 in udp_read (us=0x555555637c40, fd=6) at src/udp/udp.c:219
#12 0x00007ffff7f74d79 in udp_read_handler (flags=1, arg=0x555555637c40) at src/udp/udp.c:232
#13 0x00007ffff7f86588 in fd_poll (re=0x7ffff7fb1b80 <global_re>) at src/main/main.c:892
#14 0x00007ffff7f8688e in re_main (signalh=0x5555555779f8 <signal_handler>) at src/main/main.c:1026
#15 0x0000555555577a73 in re_main_timeout (timeout_ms=5000) at test/test.c:39
#16 0x00005555555692a7 in test_call_progress () at test/call.c:879
#17 0x0000555555577e3f in run_tests () at test/main.c:92
#18 0x000055555557827b in main (argc=1, argv=0x7fffffffe4a8) at test/main.c:249
```

**Comment by cspiel1** [comment]
Fixed the segfault. Have to investigate further, because `sdp_decode()` called in `sipsess_progr_handler()` seems to work not like expected. The audio `rdir` is still `SDP_SENDRECV` after the function returns. So the ringback is stopped by `UA_EVENT_CALL_PROGRESS`.

**Comment by cspiel1** [comment]
For the latter problem had to make a change in libre. https://github.com/baresip/re/pull/99

Tested with baresip on both sides.

**Pull request merged**